Automattic/woocommerce-payments
# Replaces upe wp_ajax calls with wc_ajax calls

Fixes #2391 

#### Changes proposed in this Pull Request
> While testing UPE with multi-currency it was discovered that the store currency was used when creating and updating the intent used by UPE. This was due to the fact that the intent requests were running through `admin-ajax` and so always reverting to the store currency instead of the customer's selected currency.
>
> The solution to this is to use `wc_ajax` requests instead.

# UPE Compatibility with Multi-Currency

Currently, UPE uses AJAX requests to create payment/setup intents to load the UPE in payment forms: the type of AJAX requests made are `wp_ajax` requests, which are sent in an admin context. In the admin context, when choosing a currency for the UPE, the function `get_woocommerce_currency` returns the store owner's currency. With Multi-Currency enabled, however, the store owner can enable multiple currencies on their store and allow customers to choose their own preferred currency. In order to use the customer's chosen currency, instead of the default site currency, we need to send these AJAX requests as `wc_ajax` requests, so that the requests are made within the client's context instead of the site admin's.

This PR modifies the UPE gateway and client side scripts, such that requests to create or update Stripe intents are sent via `wc_ajax`, so that we can correctly use the customer's chosen currency. I also updated the AJAX call to process the checkout to use `wc_ajax` instead of `wp_ajax`. That is because we can see that the checkout function is also registered in the `wc` context [over here](https://github.com/woocommerce/woocommerce/blob/trunk/includes/class-wc-ajax.php#L109-L119) and our process payment function that eventually gets called by the checkout function updates the intention [over here](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L306) with the currency taken from the order. Hopefully, this ensures that the correct currency is used at each of these junctions. 

One last thing to be aware of is that until #2418 is resolved, we are not limiting the UPE payment methods available depending on the currency chosen. This currently means that if the store has a payment method enabled that only supports a certain currency and the customer chooses another currency to use as they checkout, the UPE will throw an error. Right now this is expected behaviour, but this should get fixed in #2418, while this PR is the piece of the puzzle to ensure that the correct currency is used in that filtering. 




#### Testing instructions


1. Ensure you have pulled the changes in this PR and the latest version of the Payments Server. Also, please enable the UPE gateway using the WCPay Dev Plugin or simply set `_wcpay_feature_upe` to a value of 1 in your `wp_options` table. Finally, please enable Customer Multi-Currency using the WCPay Dev Plugin or simply set `_wcpay_feature_multi_currency` to a value of 1 in your `wp_options` table.
2. With Multi-Currency enabled, please go to **WooCommerce > Settings > Multi-currency** and ensure the currencies you would like to enable for your store are added and listed as "Enabled Currencies".
3. As a customer, if the `Automatically switch customers to the local currency if it is enabled above.` option is selected at on the Multi-Currency settings page, your local currency will be selected automatically. There are two other ways (that I am aware of) that a customer can manually select their own currency. If you are logged into a customer account, you can go to **My account > Account details** and find a dropdown to select your `Default currency` at the bottom of this page. Alternatively, there is an option to `Add a currency switcher to the cart widget` on the Multi-Currency settings page. If you select this checkbox, you can also click `Configure now` to place it where you like. For example, I added the widget to my sidebar, which allows me to change my currency at any point during the checkout process.
![Currency switcher widget on sidebar](https://cdn-std.droplr.net/files/acc_1104206/ryIjg8)
_Adding currency switcher widget to sidebar_

4. Now simply add a product to your cart and continue through a checkout with your chosen currency. The simplest way to ensure that the UPE requests are using the correct currency is to switch between USD and EUR with all payment methods enabled. You should notice that if the customer currency is set to USD, the UPE will throw an error and fail to load (because Giropay and Sofort require EUR); if the customer currency is set to EUR, the UPE should load properly and you should be able to successfully complete a checkout. There are probably better ways to confirm this, but this is what I did.
5. If you can think of some better ways to test multi-currency at checkout than the options I've offered in the previous step, now is your time to try them.



-------------------
